### PR TITLE
Mul 1795 markets api

### PR DESCRIPTION
**Pull request by Enmk**
WS Markets API:
 * get_rates
 * get_rates_history

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -28,37 +35,45 @@ var waitGroup = &sync.WaitGroup{}
...
+	//	[]string{"BTC", "ETH", "GOLOS", "BTS", "STEEM", "WAVES", "LTC", "BCH", "ETC", "DASH", "EOS"}
+	//	configuration.TargetCurrencies = []string{"LTC","DASH"}
 	configuration.TargetCurrencies = []string{"BTC", "ETH", "GOLOS", "BTS", "STEEM", "WAVES", "LTC", "BCH", "ETC", "DASH", "EOS"}
-		configuration.ReferenceCurrencies = []string{"USDT", "BTC"}
+	configuration.ReferenceCurrencies = []string{"USDT", "BTC"}
 	configuration.Exchanges = []string{"BINANCE", "BITFINEX"}
 	configuration.RefreshInterval = 1
 
-
 	configuration.HistoryApiKey = "A502B3C1-9C40-446F-9831-CA12EC039AB8"
```
Comment (vovapi) [inline cmd/main.go:48]: Нужно удостовериться, что этот ключ больше не валиден

File: currencies/currencies.go [hunk 2]
```diff
@@ -14,8 +14,6 @@ func (b *CurrencyPair) Symbol() string {
 }
 
 func (b *CurrencyPair) IsEqualTo(pair CurrencyPair) bool {
```
Comment (vovapi) [inline currencies/currencies.go:16]: maybe rename to `Equals`?

Comment (vovapi) [reply currencies/currencies.go:16]: seems that `==` can work as well

File: exchange-rates/dbManager.go [hunk 3]
```diff
@@ -14,7 +14,7 @@ import (
 	_ "github.com/lib/pq"
 )
 
-var log = slf.WithContext("exchange-rates")
+var log = slf.WithContext("DB")
```
Comment (vovapi) [inline exchange-rates/dbManager.go:17]: @jekabolt context names are lowercase?

File: exchange-rates/dbManager.go [hunk 4]
```diff
@@ -113,12 +110,13 @@ func (b *DbManager) fillRateFromSA() {
 	}
 }
 
-func (b *DbManager) getRates(timeStamp time.Time, exchangeTitle string, targetCode string, refereciesCodes []string) []DbRate {
+func (b *DbManager) getRates(from time.Time, to time.Time, exchangeTitle string, targetCode string, refereciesCodes []string) ([]DbRate, error) {
```
Comment (vovapi) [inline exchange-rates/dbManager.go:113]: rename receiver to `mgr` or `manager` or `db`?

File: exchange-rates/exchangeManager.go [hunk 5]
```diff
@@ -135,7 +135,7 @@ func (b *ExchangeManager) StartGetingData() {
 		//	}
 
 		case response := <-responseCh:
-			fmt.Printf("got history %@,  %@ %@ %s \n", response.Exchange.String(), response.Pair.TargetCurrency.CurrencyCode(), response.Pair.ReferenceCurrency.CurrencyCode(), response.OhlcvData)
+			fmt.Printf("got history %s,  %s %s %v \n", response.Exchange.String(), response.Pair.TargetCurrency.CurrencyCode(), response.Pair.ReferenceCurrency.CurrencyCode(), response.OhlcvData)
```
Comment (vovapi) [inline exchange-rates/exchangeManager.go:138]: use logger

File: exchange-rates/exchangeManager.go [hunk 6]
```diff
@@ -270,7 +270,7 @@ func (b *ExchangeManager) add(tikers server.Tickers) {
 
 			if v, ok := b.exchanges[exchangeTicker.Exchange]; ok {
 				if v.Tickers == nil {
-					v.Tickers =  map[string]Ticker{}
+					v.Tickers = map[string]Ticker{}
```
Comment (vovapi) [inline exchange-rates/exchangeManager.go:273]: initialization via struct literal is bad (for some reason)
use `make(map[string]Ticker)`

File: exchange-rates/exchangeManager.go [hunk 7]
```diff
@@ -285,38 +285,39 @@ func (b *ExchangeManager) add(tikers server.Tickers) {
...
-
-		for _, value := range historyData.OhlcvData {
-			var ticker = Ticker{}
-			ticker.TimpeStamp = value.Time_open
-			ticker.Pair = historyData.Pair
-			ticker.Rate, _ = value.Price_open.Float64()
+	if _, ok := b.exchanges[historyData.Exchange.String()]; !ok {
+		var ex = Exchange{}
+		ex.name = historyData.Exchange.String()
+		b.exchanges[historyData.Exchange.String()] = ex
```
Comment (vovapi) [inline exchange-rates/exchangeManager.go:291]:
```go
b.exchanges[historyData.Exchange.String()] = Exchange {
    name: historyData.Exchange.String()
}
```

File: cmd/main.go [hunk 8]
```diff
@@ -2,16 +2,23 @@ package main
 
 import (
 	"sync"
+	"time"
 
 	"github.com/Appscrunch/Multy-back-exchange-service/core"
 	"github.com/Appscrunch/Multy-back-exchange-service/exchange-rates"
+	"github.com/Appscrunch/Multy-back-exchange-service/server"
```
Comment (vovapi) [inline cmd/main.go:9]: move to Multy-io

File: server/socket-io.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+See LICENSE for details
+*/
+package server
+
+import (
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/graarh/golang-socketio"
```
Comment (vovapi) [inline server/socket-io.go:13]: vendor `golang-socketio`?

File: exchange-rates/dbManager.go [hunk 10]
```diff
@@ -32,7 +32,7 @@ type DbTicker struct {
 	ReferenceCurrency currencies.Currency
 	Rate              float64
 	TimpeStamp        time.Time
```
Comment (vovapi) [inline exchange-rates/dbManager.go:34]: typo

File: currencies/currencies.go (hunk 2 above)
Comment (Enmk) [reply currencies/currencies.go:16]: Nah, there are tons of things to refactor, but the main goal of this is to add WS API support. So I refrain myself from diving into rabbit hole of mass-refactoring.

File: exchange-rates/dbManager.go (hunk 4 above)
Comment (vovapi) [reply exchange-rates/dbManager.go:113]: also `referecesCodes` not `refereciesCodes`

File: exchange-rates/dbManager.go (hunk 3 above)
Comment (jekabolt) [reply exchange-rates/dbManager.go:17]: This context affect only to log prefix i.e it doesn't matter

File: .gitignore [hunk 11]
```diff
@@ -6,6 +6,9 @@
 .idea/
 exchanger
 
+# Logs of the binary
+cmd/logs/*
+
 # Test binary, build with `go test -c`
 *.test
```
Comment (jekabolt) [inline .gitignore:14]: Add exchanger.config to gitignore and add exchanger.config-example to cmd folder

File: server/socket-io.go [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+	handler.On(gosocketio.OnConnection, func(c *gosocketio.Channel) {
+		log.Debugf("Connected: %v from %v", c.Id(), c.Ip())
+	})
+
+	handler.On(gosocketio.OnDisconnection, func(c *gosocketio.Channel) {
+		log.Debugf("Disconnected: %v", c.Id())
+	})
+
+	handler.On("/get_rates", func(c *gosocketio.Channel, request marketRateRequest) marketRateResponse {
```
Comment (jekabolt) [inline server/socket-io.go:88]: use route "get:rates"

File: server/socket-io.go [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+			rates[strconv.Itoa(int(ticker.Pair.TargetCurrency))] = ticker.Rate
+		}
+
+		log.Debugf("Resulting rates count: %v", len(rates))
+		return marketRateResponse{
+			Rates: rates,
+		}
+	})
+
+	handler.On("/get_rates_history", func(c *gosocketio.Channel, request marketRateHistoryRequest) marketRateHistoryResponse {
```
Comment (jekabolt) [inline server/socket-io.go:110]: use route "get:rates:history"

**Pull request closed without merging**